submissions/nadsatt/post-oop/index.js

```diff
@@ -0,0 +1,122 @@
+/* 
+   Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: https://github.com/nadsatt/a-tiny-JS-world
+   Web app: https://nadsatt.github.io/a-tiny-JS-world/
+*/
+
+// ======== OBJECTS DEFINITIONS ========
+
+class FriendsManager {
+   constructor(friends = []){
+      this.friends = friends;
+   }
+
+   addFriends(...friends){
+      this.friends.push(...friends);
+   }
+
+   getFriends(){
+      return this.friends.map(friend => friend.name).join(' ') || 'no friends';
+   }
+}
+
+class Inhabitant {
+   constructor(species, saying, gender, name, friends){
+      this.friendsManager = new FriendsManager(friends);
+      this.species = species;
+      this.saying = saying;
+      this.gender = gender;
+      this.name = name;
+   }
+
+   toString(){
+      return Object.entries(this).slice(1).map(([key, value]) => `${key}: ${value}`).join(', ') + `, friends: ${this.friendsManager.getFriends()}`;
```

Comment:
This method now has nice optimization (`.map(....`), although with something that will fire back (`Object.entries(this)`).

The latter part presumes we'd need every property of any child class get printed. What if not? Then we'd need to patch the `Inhabitant` class instead of focusing on the particular child class.

`.slice(1)` also relies upon the fact that `friends` would always be the first property in the object, but this is not guaranteed by design and depends on JS engine in use. Also the purpose of `.slice()` is somewhat cryptic and not really clear when reading the code fragment.

Let's have an explicit list of props here. And let's give the child classes control over their properties back to them.